From 4d9c8e63e1d6b54097a3df78e21d9b62e0ae832d Mon Sep 17 00:00:00 2001
From: Huiwen He <hehuiwen@kylinos.cn>
Date: Thu, 22 Jan 2026 16:45:34 +0800
Subject: [PATCH 4/9] smb/client: refactor ntstatus_to_dos to return a pointer
 to mapping struct

Modify 'ntstatus_to_dos' to return a 'const struct ntstatus_to_dos *'
instead of populating output parameters. This allows the caller to
access all fields of the mapping entry directly.

Update 'map_smb_to_linux_error' and 'cifs_print_status' to use the
new interface.

Signed-off-by: Huiwen He <hehuiwen@kylinos.cn>
---
 fs/smb/client/smb1maperror.c | 52 +++++++++++++++++-------------------
 1 file changed, 25 insertions(+), 27 deletions(-)

diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index aba9332af291..eb8e90d2d285 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -658,42 +658,30 @@ static const struct nt_err_code_struct ntstatus_to_dos_map[] = {
 	0, 0, 0 , NULL }
 };
 
-/*****************************************************************************
- Print an error message from the status code
- *****************************************************************************/
-static void
-cifs_print_status(__u32 status_code)
+static const struct nt_err_code_struct *
+ntstatus_to_dos(__u32 ntstatus)
 {
 	int i;
 
 	for (i = 0; ntstatus_to_dos_map[i].nt_errstr != NULL; i++) {
-		if (ntstatus_to_dos_map[i].ntstatus == status_code) {
-			pr_notice("Status code returned 0x%08x %s\n",
-				  status_code, ntstatus_to_dos_map[i].nt_errstr);
-			return;
-		}
+		if (ntstatus == ntstatus_to_dos_map[i].ntstatus)
+			return &ntstatus_to_dos_map[i];
 	}
+	return NULL;
 }
 
 
+/*****************************************************************************
+ Print an error message from the status code
+ *****************************************************************************/
 static void
-ntstatus_to_dos(__u32 ntstatus, __u8 *eclass, __u16 *ecode)
+cifs_print_status(__u32 status_code)
 {
-	int i;
-	if (ntstatus == 0) {
-		*eclass = 0;
-		*ecode = 0;
-		return;
-	}
-	for (i = 0; ntstatus_to_dos_map[i].ntstatus; i++) {
-		if (ntstatus == ntstatus_to_dos_map[i].ntstatus) {
-			*eclass = ntstatus_to_dos_map[i].dos_class;
-			*ecode = ntstatus_to_dos_map[i].dos_code;
-			return;
-		}
-	}
-	*eclass = ERRHRD;
-	*ecode = ERRgeneral;
+	const struct nt_err_code_struct *map = ntstatus_to_dos(status_code);
+
+	if (map)
+		pr_notice("Status code returned 0x%08x %s\n",
+			  status_code, map->nt_errstr);
 }
 
 int
@@ -715,11 +703,21 @@ map_smb_to_linux_error(char *buf, bool logErr)
 		/* translate the newer STATUS codes to old style SMB errors
 		 * and then to POSIX errors */
 		__u32 err = le32_to_cpu(smb->Status.CifsError);
+		const struct nt_err_code_struct *map;
+
 		if (logErr && (err != (NT_STATUS_MORE_PROCESSING_REQUIRED)))
 			cifs_print_status(err);
 		else if (cifsFYI & CIFS_RC)
 			cifs_print_status(err);
-		ntstatus_to_dos(err, &smberrclass, &smberrcode);
+
+		map = ntstatus_to_dos(err);
+		if (map) {
+			smberrclass = map->dos_class;
+			smberrcode = map->dos_code;
+		} else {
+			smberrclass = ERRHRD;
+			smberrcode = ERRgeneral;
+		}
 	} else {
 		smberrclass = smb->Status.DosError.ErrorClass;
 		smberrcode = le16_to_cpu(smb->Status.DosError.Error);
-- 
2.43.0

